qemu/tests/qemu_guest_agent_update.py

```diff
@@ -0,0 +1,151 @@
+
+from virttest import env_process
+from virttest import error_context
+from virttest import utils_misc
+from virttest import data_dir
+
+from qemu.tests.qemu_guest_agent import QemuGuestAgentBasicCheckWin
+
+
+class QemuGuestAgentUpdateTest(QemuGuestAgentBasicCheckWin):
+
+    @error_context.context_aware
+    def gagent_check_pkg_update(self, test, params, env):
+
+        def copy_pkg(qemu_ga_pkg, dst_path):
+            """
+            copy qemu-ga to a specify dir and rename.
+            src_path,E:\guest-agent\qemu-ga-xx.msi
+            dst_path,C:\qemu-ga\qemu-ga-xx.msi
+            """
+            src_path = self.get_qga_pkg_path(qemu_ga_pkg, test,
+                                             session, params, self.vm)
+            session.cmd("copy /Y %s %s" % (src_path, dst_path))
+
+        def change_agent_media(cdrom_virtio):
+            """
+            change iso for virtio-win
+            :param cdrom_virtio: iso file
+            """
+            virtio_iso = utils_misc.get_path(data_dir.get_data_dir(),
+                                             cdrom_virtio)
+            logging.info("Iso is: %s" % virtio_iso)
+            self.vm.change_media("drive_virtio", virtio_iso)
+
+        def create_dst_path():
+            """
+            find the previous and the latest qemu-ga pkg and
+            copy them to a specify guest dir
+            """
+            s, o = session.cmd_status_output("mkdir %s" % self.gagent_guest_dir)
+            if s and "already exists" not in o:
+                test.error("Could not create qemu-ga directory in "
+                           "VM '%s', detail: '%s'" % (vm.name, o))
+
+            if self.gagent_src_type == "virtio-win":
+                copy_pkg(self.qemu_ga_pkg, dst_path_lat)
+                # get the previous qemu-ga pkg
+                cdrom_virtio_downgrade = params["cdrom_virtio_downgrade"]
+                error_context.context("Change cdrom to %s" % cdrom_virtio_downgrade,
+                                      logging.info)
+                change_agent_media(cdrom_virtio_downgrade)
+                time.sleep(5)
+                copy_pkg(qemu_ga_pkg_pre, dst_path_pre)
+
+            elif self.gagent_src_type == "url":
+                copy_pkg(self.qemu_ga_pkg, dst_path_lat)
+
+                # get the previous qemu-ga pkg
+                qga_html = params["qga_html"]
+                qga_url = params["qga_url"]
+                gagent_html_download_cmd = params["gagent_html_download_cmd"]
+                process.system(gagent_html_download_cmd, login_timeout)
+
+                with open(qga_html, "r") as f:
+                    lines = f.readlines()
+                list_qga = []
+                for line in lines:
+                    if "qemu-ga-win" in line:
+                        list_qga.append(line)
+                tgt_line = list_qga[-2]
+                # qemu-ga-win-7.4.4-1
+                pattern = re.compile(r"qemu-ga-win-\d\.\d\.\d-\d")
```

Comment:
more info for this comment or delete this

Reply:
I list it to make the next line more clearly